coral-incremental/src/main/java/com/linkedin/coral/incremental/RelIncrementalTransformer.java

```diff
@@ -0,0 +1,111 @@
+import org.apache.calcite.rel.RelShuttleImpl;
+import org.apache.calcite.rel.core.TableScan;
+import org.apache.calcite.rel.logical.LogicalAggregate;
+import org.apache.calcite.rel.logical.LogicalFilter;
+import org.apache.calcite.rel.logical.LogicalJoin;
+import org.apache.calcite.rel.logical.LogicalProject;
+import org.apache.calcite.rel.logical.LogicalTableScan;
+import org.apache.calcite.rel.logical.LogicalUnion;
+import org.apache.calcite.rex.RexBuilder;
+import org.apache.calcite.rex.RexNode;
+
+
+public class RelIncrementalTransformer {
+
+  private RelIncrementalTransformer() {
+  }
+
+  public static RelNode convertRelIncremental(RelNode originalNode) {
+    RelShuttle converter = new RelShuttleImpl() {
+      @Override
+      public RelNode visit(TableScan scan) {
+        RelOptTable originalTable = scan.getTable();
+        List<String> incrementalNames = new ArrayList<>(originalTable.getQualifiedName());
+        String deltaTableName = incrementalNames.remove(incrementalNames.size() - 1) + "_delta";
+        incrementalNames.add(deltaTableName);
+        RelOptTable incrementalTable =
+            RelOptTableImpl.create(originalTable.getRelOptSchema(), originalTable.getRowType(), incrementalNames, null);
+        return LogicalTableScan.create(scan.getCluster(), incrementalTable);
+      }
+
+      @Override
+      public RelNode visit(LogicalJoin join) {
+        RelNode left = join.getLeft();
+        RelNode right = join.getRight();
+        RelNode incrementalLeft = transformToIncremental(left);
+        RelNode incrementalRight = transformToIncremental(right);
+
+        RexBuilder rexBuilder = join.getCluster().getRexBuilder();
+
+        LogicalProject p1 = createProjectOverJoin(join, left, incrementalRight, rexBuilder);
+        LogicalProject p2 = createProjectOverJoin(join, incrementalLeft, right, rexBuilder);
+        LogicalProject p3 = createProjectOverJoin(join, incrementalLeft, incrementalRight, rexBuilder);
+
+        LogicalUnion unionAllJoins =
+            LogicalUnion.create(Arrays.asList(LogicalUnion.create(Arrays.asList(p1, p2), true), p3), true);
+        return unionAllJoins;
+      }
+    };
+    return originalNode.accept(converter);
+  }
+
+  private static LogicalProject createProjectOverJoin(LogicalJoin join, RelNode left, RelNode right,
+      RexBuilder rexBuilder) {
+    LogicalJoin incrementalJoin =
+        LogicalJoin.create(left, right, join.getCondition(), join.getVariablesSet(), join.getJoinType());
+    ArrayList<RexNode> projects = new ArrayList<>();
+    ArrayList<String> names = new ArrayList<>();
+    IntStream.range(0, incrementalJoin.getRowType().getFieldList().size()).forEach(i -> {
+      projects.add(rexBuilder.makeInputRef(incrementalJoin, i));
+      names.add(incrementalJoin.getRowType().getFieldNames().get(i));
+    });
+    return LogicalProject.create(incrementalJoin, projects, names);
+  }
+
+  private static RelNode transformToIncremental(RelNode target) {
```

Comment:
When we discussed this, we assumed the simple case where it goes up to the leaf and converts it to `table_delta`. I see you have implemented a more complex use case where it recursively calls `convertRelIncremental`. A few questions:
1- Have we verified the rewrites are correct? Let us write a short doc to illustrate with an example, or may be in the comments.
2- Since it is recursive with `convertRelIncremental`, can we move the logic to the visitors above?

Reply:
1. Yes, `RelToIncrementalSqlConverterTest` has a number of tests (with inputs and expected outputs) to verify the correctness of the rewrites. I'm happy to throw these into a doc as well if that would be more digestible.

2. Good point, will refactor this.